drop-storage/src/lib.rs

```diff
@@ -552,49 +546,62 @@ impl Storage {
-            .map_err(error::Error::DBError)?
-            .iter()
-            .map(|s| TransferCancelState {
-                transfer_id: transfer.id,
-                by_peer: s.by_peer != 0,
-                created_at: s.created_at.timestamp_millis(),
-            })
-            .collect();
-
-            transfer.failed_states = sqlx::query!(
-                "SELECT status_code, created_at FROM transfer_failed_states WHERE transfer_id = ?1",
-                tid
-            )
-            .fetch_all(&mut *conn)
-            .await
-            .map_err(error::Error::DBError)?
-            .iter()
-            .map(|s| TransferFailedState {
-                transfer_id: transfer.id,
-                status_code: s.status_code,
-                created_at: s.created_at.timestamp_millis(),
-            })
-            .collect();
+            transfer.states.extend(
+                sqlx::query!(
+                    "SELECT created_at  FROM transfer_active_states WHERE transfer_id = ?1",
+                    tid
+                )
+                .fetch_all(&mut *conn)
+                .await
+                .map_err(error::Error::DBError)?
+                .iter()
+                .map(|s| TransferStateEvent {
+                    transfer_id: transfer.id,
+                    created_at: s.created_at.timestamp_millis(),
+                    data: types::TransferStateEventData::Active,
+                }),
+            );
+
+            transfer.states.extend(
+                sqlx::query!(
+                    "SELECT by_peer, created_at FROM transfer_cancel_states WHERE transfer_id = ?1",
+                    tid
+                )
+                .fetch_all(&mut *conn)
+                .await
+                .map_err(error::Error::DBError)?
+                .iter()
+                .map(|s| TransferStateEvent {
+                    transfer_id: transfer.id,
+                    created_at: s.created_at.timestamp_millis(),
+                    data: types::TransferStateEventData::Cancel {
+                        by_peer: s.by_peer != 0,
+                    },
+                }),
+            );
+
+            transfer.states.extend(
+                sqlx::query!(
+                    "SELECT status_code, created_at FROM transfer_failed_states WHERE transfer_id \
+                     = ?1",
+                    tid
+                )
+                .fetch_all(&mut *conn)
+                .await
+                .map_err(error::Error::DBError)?
+                .iter()
+                .map(|s| TransferStateEvent {
+                    transfer_id: transfer.id,
+                    created_at: s.created_at.timestamp_millis(),
+                    data: types::TransferStateEventData::Failed {
+                        status_code: s.status_code,
+                    },
+                }),
+            );
+
+            transfer
+                .states
+                .sort_by(|a, b| a.created_at.cmp(&b.created_at));
```

Comment:
The sqlx gives you the `NaiveDateTime` which has a better precision than `timestamp_millis()` so maybe we should sort by the original thing and then convert it to ms or just use the conversions during the serialization with `#[serde(serialize_with = "NaiveDateTime::timestamp_millis()")]`

Reply:
If that's the case then there is no point in keeping the `NaiveDateTime`. But still, I think that using the sqlx native type would be beneficial because then we can then use the `sqlx::query_as!()` macro with the `types.rs` structures directly.
I'm leaving the suggestion for you to decide

Reply:
I changed the fields to be `NaiveDateTime`, though it seems to use `query_as!` there would need to be quite a big change to the queries, ill come back to that in a different MR in the future, since it's not a pressing concern right now